sysl2/sysl/validate.go

```diff
@@ -0,0 +1,355 @@
+			} else if ref1.GetPath() != nil && ref2.GetPath() != nil {
+				expectedType = ref1.GetPath()[0]
+				return ref1.GetPath()[0] == ref2.GetPath()[0], expectedType
+			}
+		case type1.GetTypeRef().GetRef().GetAppname() != nil:
+			expectedType = type1.GetTypeRef().GetRef().GetAppname().GetPart()[0]
+		case type1.GetTypeRef().GetRef().GetPath() != nil:
+			expectedType = type1.GetTypeRef().GetRef().GetPath()[0]
+		}
+	}
+
+	return false, expectedType
+}
+
+func resolveVariableType(expr *sysl.Expr, viewName string) (*sysl.Type, []ValidationMsg) {
+	validationMsgs := make([]ValidationMsg, 0, 10)
+
+	switch expr.Expr.(type) {
+	case *sysl.Expr_Transform_:
+		if typeRef := expr.GetType().GetTypeRef(); typeRef != nil {
+			return &sysl.Type{
+				Type: &sysl.Type_TypeRef{
+					TypeRef: &sysl.ScopedRef{
+						Ref: &sysl.Scope{Path: typeRef.GetRef().GetAppname().GetPart()},
+					},
+				},
+			}, validationMsgs
+		}
+		return expr.GetType(), validationMsgs
+	case *sysl.Expr_Literal:
+		return expr.GetType(), validationMsgs
+	case *sysl.Expr_Unexpr:
+		varType, messages := resolveVariableType(expr.GetUnexpr().GetArg(), viewName)
+		validationMsgs = append(validationMsgs, messages...)
+		if isSame, expected := hasSameType(varType, boolType); !isSame {
+			validationMsgs = append(validationMsgs, ValidationMsg{
+				Message: fmt.Sprintf("In view '%s', unary operator used with non boolean type: '%s'", viewName, expected),
+				MsgType: ERROR})
+		}
+		return boolType, validationMsgs
+	}
+
+	return noType(), validationMsgs
+}
+
+func validateTypes(specModule *sysl.Module, expression *sysl.Expr,
+	viewName string, implViews map[string]*sysl.View, out string) []ValidationMsg {
+	validationMsgs := make([]ValidationMsg, 0, 50)
+
+	attrDefs := map[string]*sysl.Type{}
+
+	for _, stmt := range expression.GetTransform().GetStmt() {
+		if stmt.GetAssign() != nil {
+			typeName := stmt.GetAssign().GetName()
+
+			expr := stmt.GetAssign().GetExpr()
+			attrType, messages := resolveVariableType(expr, viewName)
+			attrDefs[typeName] = attrType
+			validationMsgs = append(validationMsgs, messages...)
+
+			if expr.GetTransform() != nil {
+				attrTypeName, _ := viewOutput(attrType)
+				validationMsgs = append(validationMsgs, validateTypes(specModule, expr, viewName, implViews, attrTypeName)...)
+			}
+		}
+	}
+
+	for _, grammarSysl := range specModule.GetApps() {
+		if grammarType, exists := grammarSysl.Types[out]; exists {
+			validationMsgs = append(
+				validationMsgs,
+				compareSpecVsImpl(grammarSysl.Types, grammarType.GetTuple().GetAttrDefs(),
+					attrDefs, getImplAttrNames(attrDefs), viewName)...)
+		}
+	}
+
+	return validationMsgs
+}
+
+func validateTransform(rootTransform, transform, grammar, start string) []ValidationMsg {
+	validationMsgs := make([]ValidationMsg, 0, 100)
+
+	tx, _ := loadAndGetDefaultApp(rootTransform, transform)
+
+	grammarSyslFile := grammar[:len(grammar)-1] + "sysl"
+	grammarSysl, _ := loadAndGetDefaultApp("", grammarSyslFile)
```

Comment:
`validateTransform` is doing two things, move loading of grammar etc out of this function.

Reply:
Changed.